libs/server-sdk/src/data_store/persistent/expiration_tracker.cpp

```diff
@@ -0,0 +1,152 @@
+#include "expiration_tracker.hpp"
+
+namespace launchdarkly::server_side::data_store::persistent {
+
+void ExpirationTracker::Add(std::string const& key,
+                            ExpirationTracker::TimePoint expiration) {
+    unscoped_.insert({key, expiration});
+}
+
+void ExpirationTracker::Remove(std::string const& key) {
+    unscoped_.erase(key);
+}
+
+ExpirationTracker::TrackState ExpirationTracker::State(
+    std::string const& key,
+    ExpirationTracker::TimePoint current_time) const {
+    auto item = unscoped_.find(key);
+    if (item != unscoped_.end()) {
+        return State(item->second, current_time);
+    }
+
+    return ExpirationTracker::TrackState::kNotTracked;
+}
+
+void ExpirationTracker::Add(data_store::DataKind kind,
+                            std::string const& key,
+                            ExpirationTracker::TimePoint expiration) {
+    scoped_.Set(kind, key, expiration);
+}
+
+void ExpirationTracker::Remove(data_store::DataKind kind,
+                               std::string const& key) {
+    scoped_.Remove(kind, key);
+}
+
+ExpirationTracker::TrackState ExpirationTracker::State(
+    data_store::DataKind kind,
+    std::string const& key,
+    ExpirationTracker::TimePoint current_time) const {
+    auto expiration = scoped_.Get(kind, key);
+    if (expiration.has_value()) {
+        return State(expiration.value(), current_time);
+    }
+    return ExpirationTracker::TrackState::kNotTracked;
+}
+
+void ExpirationTracker::Clear() {
+    scoped_.Clear();
+    unscoped_.clear();
+}
+std::vector<std::pair<std::optional<DataKind>, std::string>>
+ExpirationTracker::Prune(ExpirationTracker::TimePoint current_time) {
+    std::vector<std::pair<std::optional<DataKind>, std::string>> pruned;
+
+    // Determine everything to be pruned.
+    for (auto const& item : unscoped_) {
+        if (State(item.second, current_time) ==
+            ExpirationTracker::TrackState::kStale) {
```

Comment:
I think both of these `push_back({arg1, arg2})` can be replaced with `emplace_back(arg1, arg2)` to avoid constructing a temporary